twitter_timeline/main.py

```diff
@@ -22,25 +24,149 @@ def before_request():
     g.db = connect_db(settings.DATABASE_NAME)
 
 
+# Views go here:
 @app.route('/friendship', methods=['POST', 'DELETE'])
 @json_only
 @auth_only
 def friendship(user_id):
-    pass
+    '''
+    Creates or deletes a "following relationship" between an authenticated
+    user and another user whose username is sent in a POST json object.
+    '''
+    
+    # Verify a username was sent
+    try:
+        followed_username = request.json['username']
+    except:
+        abort(400)
+    
+    # Verify that user exists
+    followed_user = g.db.users.find_one({'username': followed_username})
+    if not followed_user:
+        abort(400)
+    
+    # Get the follower object from the users collection
+    follower = g.db.users.find_one({'_id': ObjectId(user_id)})
+    
+    # You can't follow yourself.  That's just silly.
+    if followed_user['_id'] == follower['_id']:
+        abort(400)
+    
+    # Get the list of users already being followed.  This should be initialized
+    #  as a blank list if the 'following' key does not exist.
+    follower.setdefault('following', [])
+    following = follower['following']
+    # # One-liner for above
+    # following = follower['following'] if 'following' in follower else []
+    
+    if request.method == 'POST':
+        # Append the followed user's id to the follower list if it is not 
+        #  already in it.  If already following, return a Bad Request response.
+        if followed_user['_id'] not in following:
+            following.append(followed_user['_id'])
+            response = Response(status=201, mimetype=JSON_MIME_TYPE)
+        else:
+            abort(400)
+    
+    elif request.method == 'DELETE':
+        # Remove the followed user's id from the follower list if it is already
+        #  in it.  If not already following, return a Bad Request response.
+        if followed_user['_id'] in following:
+            following.remove(followed_user['_id'])
+            response = Response(status=204, mimetype=JSON_MIME_TYPE)
+        else:
+            abort(400)
+        
+    # Update the users collection with the new or revised following list
+    g.db.users.update_one(
+        {'_id': ObjectId(user_id)},
+        {'$set': {'following': following}}
+    )
+    
+    # If we successfully reach this point, return the appropriate response
+    return response
 
 
 @app.route('/followers', methods=['GET'])
 @auth_only
 def followers(user_id):
-    pass
+    '''
+    Returns list of users following the authenticated user.
+    '''
+    # The list of users to be returned
+    followers = []
+    
+    # Get all users who have the authenticated user in their 'following' list
+    for follower in g.db.users.find({ 'following': {'$in': [user_id]}}):
```

Comment:
Very nice, I previously didn't know about using $in to check a value within a list in the collection field